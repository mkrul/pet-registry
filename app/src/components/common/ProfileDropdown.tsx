import NavLink from "../shared/NavLink";
import { useNavigate } from "react-router-dom";

const ProfileDropdown = () => {
  const navigate = useNavigate();

  const handleLogout = async () => {
    try {
      navigate("/login");
    } catch (err) {
      console.error("Logout failed:", err);
    }
  };

  return (
    <div className="flex-none gap-2">
      <div className="dropdown dropdown-end">
        <div tabIndex={0} role="button" className="btn btn-ghost btn-circle avatar">
          <div className="w-10 rounded-full">
            <img
              alt="Tailwind CSS Navbar component"
              src="https://img.daisyui.com/images/stock/photo-1534528741775-53994a69daeb.webp"
            />
          </div>
        </div>
        <ul
          tabIndex={0}
          className="menu menu-sm dropdown-content bg-base-100 rounded-box z-[1] mt-3 w-52 p-2 shadow"
        >
          <li>
            <NavLink>My Reports</NavLink>
          </li>
          <li>
            <NavLink>My Pets</NavLink>
          </li>
          <li>
            <NavLink>
              Profile <span className="badge">New</span>
            </NavLink>
          </li>
          <li>
            <NavLink>Settings</NavLink>
          </li>
          <li>
            <NavLink handler={handleLogout}>Logout</NavLink>
          </li>
        </ul>
      </div>
    </div>
  );
};

export default ProfileDropdown;
