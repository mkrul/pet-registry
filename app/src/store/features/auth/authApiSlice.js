import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { setUser, clearUser, setLoading, setError, updateLastActivity } from "./authSlice";
import { addNotification } from "../notifications/notificationsSlice";

export const authApiSlice = createApi({
  reducerPath: "authApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "/api",
    credentials: "include",
    prepareHeaders: headers => {
      const token = document.querySelector('meta[name="csrf-token"]')?.getAttribute("content");
      if (token) {
        headers.set("X-CSRF-Token", token);
      }
      headers.set("Accept", "application/json");
      headers.set("Content-Type", "application/json");
      return headers;
    }
  }),
  tagTypes: ["Auth"],
  endpoints: builder => ({
    login: builder.mutation({
      query: credentials => ({
        url: "/login",
        method: "POST",
        body: credentials,
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Logged in successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.data?.error || "Login failed";
          dispatch(setError(errorMessage));
          dispatch(clearUser());
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    logout: builder.mutation({
      query: () => ({
        url: "/logout",
        method: "DELETE",
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          await queryFulfilled;
          localStorage.clear();
          sessionStorage.clear();
          dispatch(clearUser());
          dispatch(authApiSlice.util.resetApiState());
          window.location.href = "/login";
        } catch (err) {
          dispatch(setError("Logout failed"));
        }
      }
    }),
    getCurrentUser: builder.query({
      query: () => ({
        url: "/current_user",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
        } catch (err) {
          if (err?.status === 401) {
            dispatch(clearUser());
          }
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    signUp: builder.mutation({
      query: credentials => ({
        url: "/sign_up",
        method: "POST",
        body: credentials
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(setError(err?.data?.message || "Sign up failed"));
          dispatch(clearUser());
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    pollSession: builder.mutation({
      query: () => ({
        url: "/api/sessions/poll",
        method: "POST",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch }) {
        dispatch(updateLastActivity());
      }
    }),
    updateUser: builder.mutation({
      query: (userData) => ({
        url: "/profile",
        method: "PATCH",
        body: { user: userData },
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Profile updated successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.error?.data?.message || "Update failed";
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
          throw err;
        }
      }
    }),
    changePassword: builder.mutation({
      query: (passwordData) => ({
        url: "/change_password",
        method: "PATCH",
        body: { user: passwordData },
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Password changed successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.error?.data?.message || "Password change failed";
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
          throw err;
        }
      }
    })
  })
});

export const {
  useLoginMutation,
  useLogoutMutation,
  useGetCurrentUserQuery,
  useSignUpMutation,
  usePollSessionMutation,
  useUpdateUserMutation,
  useChangePasswordMutation
} = authApiSlice;
