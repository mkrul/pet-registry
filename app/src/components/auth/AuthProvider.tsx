import React, { useEffect } from "react";
import { useGetCurrentUserQuery } from "../../redux/features/auth/authApiSlice";
import { useAppDispatch } from "../../redux/hooks";
import {
  setUser,
  clearUser,
  setAuthLoading,
  checkAuthTimeout
} from "../../redux/features/auth/authSlice";

interface AuthProviderProps {
  children: React.ReactNode;
}

const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const dispatch = useAppDispatch();

  // Fetch current user with polling enabled
  const { data, isLoading } = useGetCurrentUserQuery(undefined, {
    // Poll every 5 minutes to keep session alive
    pollingInterval: 5 * 60 * 1000
  });

  // Initial auth check and periodic updates
  useEffect(() => {
    if (!isLoading) {
      if (data?.user) {
        dispatch(setUser(data.user));
      } else {
        dispatch(clearUser());
      }
    } else {
      dispatch(setAuthLoading());
    }
  }, [data, isLoading, dispatch]);

  // Check auth timeout periodically
  useEffect(() => {
    const checkTimeout = () => {
      dispatch(checkAuthTimeout());
    };

    // Check every minute
    const interval = setInterval(checkTimeout, 60 * 1000);

    // Check when tab becomes visible
    const handleVisibilityChange = () => {
      if (document.visibilityState === "visible") {
        checkTimeout();
      }
    };

    document.addEventListener("visibilitychange", handleVisibilityChange);

    // Initial check
    checkTimeout();

    return () => {
      clearInterval(interval);
      document.removeEventListener("visibilitychange", handleVisibilityChange);
    };
  }, [dispatch]);

  // Render children without additional wrapper to avoid unnecessary DOM nesting
  return <>{children}</>;
};

export default AuthProvider;
