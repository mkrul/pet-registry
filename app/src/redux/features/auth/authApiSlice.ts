import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { setUser, clearUser } from "./authSlice";
import { AuthResponse, SignUpRequest } from "../../../types/auth/AuthApiSlice";

export const authApiSlice = createApi({
  reducerPath: "authApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "/api",
    credentials: "include",
    prepareHeaders: headers => {
      headers.set("Accept", "application/json");
      headers.set("Content-Type", "application/json");
      return headers;
    }
  }),
  tagTypes: ["Auth"],
  keepUnusedDataFor: 3600,
  endpoints: builder => ({
    login: builder.mutation<AuthResponse, { user: { email: string; password: string } }>({
      query: credentials => ({
        url: "/login",
        method: "POST",
        body: credentials
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    logout: builder.mutation<{ message: string }, void>({
      query: () => ({
        url: "/logout",
        method: "DELETE"
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        try {
          await queryFulfilled;
          dispatch(clearUser());
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    getCurrentUser: builder.query<AuthResponse, void>({
      query: () => ({
        url: "/current_user",
        method: "GET"
      }),
      keepUnusedDataFor: 3600,
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          } else {
            dispatch(clearUser());
          }
        } catch (err) {
          dispatch(clearUser());
        }
      },
      providesTags: ["Auth"],
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    signUp: builder.mutation<AuthResponse, SignUpRequest>({
      query: credentials => ({
        url: "/auth/registration",
        method: "POST",
        body: credentials
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    })
  })
});

export const { useLoginMutation, useLogoutMutation, useGetCurrentUserQuery, useSignUpMutation } =
  authApiSlice;
