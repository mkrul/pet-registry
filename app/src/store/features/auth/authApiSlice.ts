import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { setUser, clearUser, setLoading, setError, updateLastActivity } from "./authSlice";
import { setNotification } from "../notifications/notificationsSlice";
import { NotificationType } from "../../../shared/types/common/Notification";
import { AuthResponse, SignUpRequest } from "../../../shared/types/auth/AuthApiSlice";

export const authApiSlice = createApi({
  reducerPath: "authApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "/api",
    credentials: "include",
    prepareHeaders: headers => {
      const token = document.querySelector('meta[name="csrf-token"]')?.getAttribute("content");
      if (token) {
        headers.set("X-CSRF-Token", token);
      }
      headers.set("Accept", "application/json");
      headers.set("Content-Type", "application/json");
      return headers;
    }
  }),
  tagTypes: ["Auth"],
  endpoints: builder => ({
    login: builder.mutation<AuthResponse, { user: { email: string; password: string } }>({
      query: credentials => ({
        url: "/login",
        method: "POST",
        body: credentials,
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
          dispatch(
            setNotification({
              type: NotificationType.SUCCESS,
              message: data.message || "Logged in successfully"
            })
          );
        } catch (err: any) {
          const errorMessage = err?.error?.data?.error || err?.data?.error || "Login failed";
          dispatch(setError(errorMessage));
          dispatch(clearUser());
          dispatch(
            setNotification({
              type: NotificationType.ERROR,
              message: errorMessage
            })
          );
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    logout: builder.mutation({
      query: () => ({
        url: "/logout",
        method: "DELETE",
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          await queryFulfilled;
          localStorage.clear();
          sessionStorage.clear();
          dispatch(clearUser());
          dispatch(authApiSlice.util.resetApiState());
          window.location.href = "/login";
        } catch (err: any) {
          dispatch(setError("Logout failed"));
        }
      }
    }),
    getCurrentUser: builder.query<AuthResponse, void>({
      query: () => ({
        url: "/current_user",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
        } catch (err: any) {
          if (err?.status === 401) {
            dispatch(clearUser());
          }
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    signUp: builder.mutation<AuthResponse, SignUpRequest>({
      query: credentials => ({
        url: "/sign_up",
        method: "POST",
        body: credentials
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err: any) {
          dispatch(setError(err?.data?.message || "Sign up failed"));
          dispatch(clearUser());
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    pollSession: builder.mutation<void, void>({
      query: () => ({
        url: "/api/sessions/poll",
        method: "POST",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch }) {
        dispatch(updateLastActivity());
      }
    })
  })
});

export const {
  useLoginMutation,
  useLogoutMutation,
  useGetCurrentUserQuery,
  useSignUpMutation,
  usePollSessionMutation
} = authApiSlice;
