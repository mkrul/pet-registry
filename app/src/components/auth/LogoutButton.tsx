import React from "react";
import { useLogoutMutation } from "../../redux/features/auth/authApiSlice";
import { useAppDispatch } from "../../redux/hooks";
import { clearUser } from "../../redux/features/auth/authSlice";

const LogoutButton: React.FC = () => {
  const dispatch = useAppDispatch();
  const [logout] = useLogoutMutation();

  const handleLogout = async () => {
    try {
      const response = await logout().unwrap();
      console.log(response.message);
      dispatch(clearUser());
      // Redirect to login if needed
    } catch (err) {
      console.error("Logout failed:", err);
    }
  };

  return <button onClick={handleLogout}>Logout</button>;
};

export default LogoutButton;
