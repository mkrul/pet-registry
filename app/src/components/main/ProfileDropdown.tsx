import { useRef, useState, useEffect } from "react";
import NavLink from "../common/NavLink";
import { useNavigate } from "react-router-dom";
import LogoutButton from "../auth/LogoutButton";
import { useSelector } from "react-redux";
import { RootState } from "../../redux/store";
import { useGetCurrentUserQuery, useLogoutMutation } from "../../redux/features/auth/authApiSlice";

const ProfileDropdown: React.FC = () => {
  const navigate = useNavigate();
  const [isOpen, setIsOpen] = useState(false);
  const [isImageLoaded, setIsImageLoaded] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const isAuthenticated = useSelector((state: RootState) => !!state.auth.user);
  const { isLoading } = useGetCurrentUserQuery();

  useEffect(() => {
    const img = new Image();
    img.src = "/images/placeholder-blue.jpg";
    img.onload = () => setIsImageLoaded(true);
  }, []);

  const [logout] = useLogoutMutation();

  const handleToggle = () => {
    if (!isAuthenticated) {
      navigate("/login");
      return;
    }
    setIsOpen(!isOpen);
  };

  const handleLogout = async () => {
    try {
      await logout({}).unwrap();
      window.location.href = "/";
    } catch (err) {
      console.error("Failed to logout:", err);
    }
  };

  if (isLoading || !isImageLoaded) {
    return (
      <div className="flex-none">
        <div className="w-10 h-10 rounded-full bg-base-200 animate-pulse"></div>
      </div>
    );
  }

  if (!isAuthenticated) {
    return (
      <div className="flex-none">
        <ul className="menu menu-horizontal px-1">
          <li>
            <NavLink linkTo="/login" data-testid="nav-link-login">
              Log In
            </NavLink>
          </li>
        </ul>
      </div>
    );
  }

  return (
    <div className="flex-none gap-2">
      <div className="dropdown dropdown-end" ref={dropdownRef}>
        <div
          tabIndex={0}
          role="button"
          className="btn btn-ghost hover:bg-blue-300 btn-circle avatar"
          onClick={handleToggle}
          data-testid="profile-button"
        >
          <div className="w-10 rounded-full">
            <img alt="User avatar" src="/images/placeholder-blue.jpg" className="object-cover" />
          </div>
        </div>
        <ul
          tabIndex={0}
          className={`menu menu-sm dropdown-content bg-white rounded-box z-[1] mt-3 w-52 p-2 shadow ${
            isOpen ? "block" : "hidden"
          }`}
        >
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <NavLink>My Reports</NavLink>
          </li>
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <NavLink>My Pets</NavLink>
          </li>
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <NavLink>Profile</NavLink>
          </li>
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <NavLink>Settings</NavLink>
          </li>
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <LogoutButton onCompleted={handleLogout} />
          </li>
        </ul>
      </div>
    </div>
  );
};

export default ProfileDropdown;
