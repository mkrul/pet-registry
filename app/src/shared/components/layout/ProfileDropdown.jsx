import { useRef, useState, useEffect } from "react";
import NavLink from "../common/NavLink";
import { useNavigate } from "react-router-dom";
import LogoutButton from "../../../features/auth/components/LogoutButton.jsx";
import { useSelector } from "react-redux";
import { useGetCurrentUserQuery, useLogoutMutation } from "../../../store/features/auth/authApiSlice.js";

const ProfileDropdown = () => {
  const navigate = useNavigate();
  const [isOpen, setIsOpen] = useState(false);
  const [isImageLoaded, setIsImageLoaded] = useState(false);
  const dropdownRef = useRef(null);
  const isAuthenticated = useSelector((state) => !!state.auth.user);
  const { isLoading } = useGetCurrentUserQuery();

  useEffect(() => {
    const img = new Image();
    img.src = "/images/placeholder-blue.jpg";
    img.onload = () => setIsImageLoaded(true);
  }, []);

  const [logout] = useLogoutMutation();

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target)) {
        setIsOpen(false);
      }
    };

    if (isOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isOpen]);

  const handleToggle = () => {
    if (!isAuthenticated) {
      navigate("/login");
      return;
    }
    setIsOpen(!isOpen);
  };

  const handleLogout = async () => {
    try {
      await logout({}).unwrap();
    } catch (err) {
      console.error("Failed to logout:", err);
      window.location.href = "/login";
    }
  };

  if (isLoading || !isImageLoaded) {
    return (
      <div className="flex-none">
        <div className="w-11 h-11 rounded-full bg-base-200 animate-pulse"></div>
      </div>
    );
  }

  if (!isAuthenticated) {
    return null;
  }

  return (
    <div className="flex-none gap-2">
      <div className="dropdown dropdown-end" ref={dropdownRef}>
        <button
          className="btn btn-ghost hover:bg-blue-300 btn-circle avatar sm:mr-2"
          onClick={handleToggle}
          data-testid="profile-button"
        >
          <div className="w-11 h-11 rounded-full">
            <img alt="User avatar" src="/images/placeholder-blue.jpg" className="object-cover w-full h-full" />
          </div>
        </button>
        <ul
          className={`menu menu-sm dropdown-content bg-white rounded-box z-[1] mt-3 w-52 p-2 shadow ${
            isOpen ? "block" : "hidden"
          }`}
        >
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <NavLink linkTo="/dashboard">Dashboard</NavLink>
          </li>
          <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
            <LogoutButton onCompleted={handleLogout} />
          </li>
        </ul>
      </div>
    </div>
  );
};

export default ProfileDropdown;
