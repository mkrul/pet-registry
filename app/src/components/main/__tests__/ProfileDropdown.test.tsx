import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import ProfileDropdown from "../ProfileDropdown";
import { authApiSlice } from "../../../redux/features/auth/authApiSlice";

const mockNavigate = vi.fn();

// Mock react-router-dom
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate
  };
});

// Mock the LogoutButton component
vi.mock("../../auth/LogoutButton", () => ({
  default: ({ onCompleted }: { onCompleted: () => void }) => (
    <button data-testid="logout-button" onClick={onCompleted}>
      Logout
    </button>
  )
}));

// Mock NavLink component
vi.mock("../../common/NavLink", () => ({
  default: ({
    children,
    "data-testid": dataTestId,
    linkTo
  }: {
    children: React.ReactNode;
    "data-testid"?: string;
    linkTo?: string;
  }) => (
    <a data-testid={dataTestId || "nav-link"} onClick={() => linkTo && mockNavigate(linkTo)}>
      {children}
    </a>
  )
}));

const mockUser = {
  email: "test@example.com",
  id: 1
};

const createMockStore = (isAuthenticated: boolean) => {
  return configureStore({
    reducer: {
      auth: () => ({
        user: isAuthenticated ? mockUser : null
      }),
      [authApiSlice.reducerPath]: authApiSlice.reducer
    },
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(authApiSlice.middleware)
  });
};

const renderProfileDropdown = (isAuthenticated: boolean = true) => {
  const store = createMockStore(isAuthenticated);
  return render(
    <Provider store={store}>
      <BrowserRouter>
        <ProfileDropdown />
      </BrowserRouter>
    </Provider>
  );
};

describe("ProfileDropdown", () => {
  beforeEach(() => {
    mockNavigate.mockClear();
  });

  describe("Authenticated State", () => {
    it("renders the profile avatar button when authenticated", () => {
      renderProfileDropdown(true);
      const avatarButton = screen.getByTestId("profile-button");
      expect(avatarButton).toBeInTheDocument();
    });

    it("toggles dropdown visibility when clicked", async () => {
      renderProfileDropdown(true);
      const avatarButton = screen.getByTestId("profile-button");
      const dropdown = screen.getByRole("list");
      expect(dropdown).toHaveClass("hidden");

      fireEvent.click(avatarButton);
      expect(dropdown).toHaveClass("block");

      fireEvent.click(avatarButton);
      expect(dropdown).toHaveClass("hidden");
    });

    it.each([["My Reports"], ["My Pets"], ["Profile"], ["Settings"]])(
      "displays %s link in dropdown when authenticated",
      async linkText => {
        renderProfileDropdown(true);
        const avatarButton = screen.getByTestId("profile-button");
        fireEvent.click(avatarButton);

        await waitFor(() => {
          const links = screen.getAllByTestId("nav-link");
          expect(links.some(link => link.textContent === linkText)).toBe(true);
        });
      }
    );
  });

  describe("Unauthenticated State", () => {
    it("renders login button when not authenticated", () => {
      renderProfileDropdown(false);
      const loginButton = screen.getByTestId("nav-link-login");
      expect(loginButton).toBeInTheDocument();
      expect(loginButton).toHaveTextContent("Log In");
    });

    it("navigates to login page when login button is clicked", () => {
      renderProfileDropdown(false);
      const loginButton = screen.getByTestId("nav-link-login");
      fireEvent.click(loginButton);
      expect(mockNavigate).toHaveBeenCalledWith("/login");
    });
  });

  describe("Logout Functionality", () => {
    it("navigates to login page after logout", async () => {
      renderProfileDropdown(true);
      const avatarButton = screen.getByTestId("profile-button");
      fireEvent.click(avatarButton);
      fireEvent.click(screen.getByTestId("logout-button"));

      await waitFor(() => {
        expect(mockNavigate).toHaveBeenCalledWith("/login");
      });
    });
  });
});
