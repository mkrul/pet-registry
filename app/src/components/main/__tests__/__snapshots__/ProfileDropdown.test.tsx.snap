// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileDropdown > Rendering > renders the profile avatar button 1`] = `
<div>
  <div
    class="flex-none gap-2"
  >
    <div
      class="dropdown dropdown-end"
    >
      <div
        class="btn btn-ghost btn-circle avatar"
        role="button"
        tabindex="0"
      >
        <div
          class="w-10 rounded-full"
        >
          <img
            alt="Tailwind CSS Navbar component"
            src="https://img.daisyui.com/images/stock/photo-1534528741775-53994a69daeb.webp"
          />
        </div>
      </div>
      <ul
        class="menu menu-sm dropdown-content bg-base-300 rounded-box z-[1] mt-3 w-52 p-2 shadow hidden"
        tabindex="0"
      >
        <li
          class="hover:bg-base-100 rounded-lg transition-colors duration-200"
        >
          <a
            data-testid="nav-link"
          >
            My Reports
          </a>
        </li>
        <li
          class="hover:bg-base-100 rounded-lg transition-colors duration-200"
        >
          <a
            data-testid="nav-link"
          >
            My Pets
          </a>
        </li>
        <li
          class="hover:bg-base-100 rounded-lg transition-colors duration-200"
        >
          <a
            data-testid="nav-link"
          >
            Profile
          </a>
        </li>
        <li
          class="hover:bg-base-100 rounded-lg transition-colors duration-200"
        >
          <a
            data-testid="nav-link"
          >
            Settings
          </a>
        </li>
        <li
          class="hover:bg-base-100 rounded-lg transition-colors duration-200"
        >
          <button
            data-testid="logout-button"
          >
            Logout
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
